File: src/javascript/node/sessionless.js
import { v4 as uuidv4 } from 'uuid';
import { secp256k1 } from 'ethereum-cryptography/secp256k1';
import { keccak256 } from "ethereum-cryptography/keccak.js";
import { bytesToHex } from "ethereum-cryptography/utils.js";
import { utf8ToBytes } from "ethereum-cryptography/utils.js";

let getKeysFromDisk;

const AsyncFunction = (async () => {}).constructor;

const generateKeys = async (saveKeys, getKeys) => {
  if(!saveKeys || !getKeys) {
    throw new Error(`Since this can be run on any machine with node, there is no default secure storage. You will need to provide a saveKeys and getKeys function`);
  }
  const privateKey = bytesToHex(secp256k1.utils.randomPrivateKey());
  const pubKey = bytesToHex(secp256k1.getPublicKey(privateKey));
  saveKeys && (saveKeys instanceof AsyncFunction ? await saveKeys({
    privateKey,
    pubKey
  }) : saveKeys({
    privateKey,
    pubKey
  }));
  getKeysFromDisk = getKeys;
  return {
    privateKey,
    pubKey
  };
};

const getKeys = async () => {
  if(!getKeysFromDisk) {
    throw new Error(`Since this can be run on any machine with node, there is no default secure storage. You will need to have your own getKeys function.`);
  } else {
    return getKeysFromDisk instanceof AsyncFunction ? await getKeysFromDisk() : getKeysFromDisk();
  }
};

const sign = async (message) => {
  const { privateKey } = await getKeys();
  const messageHash = keccak256(utf8ToBytes(message)).slice(32);
  const signatureAsBigInts = secp256k1.sign(messageHash, privateKey);
  const signature = signatureAsBigInts.r.toString(16) + signatureAsBigInts.s.toString(16);
  return signature;
};

const verifySignature = (sig, message, pubKey) => {
  const messageHash = keccak256(utf8ToBytes(message)).slice(32);
  
  let signature = {
    r: sig.substring(0, 64),
    s: sig.substring(64)
  };

  let hex = signature.r;
  if (hex.length % 2) { 
    hex = '0' + hex; 
  }

  const bn = BigInt('0x' + hex);

  signature.r = bn;

  let hex2 = signature.s;
  if (hex2.length % 2) { 
    hex2 = '0' + hex2; 
  }

  const bn2 = BigInt('0x' + hex2);

  signature.r = bn;
  signature.s = bn2;

  const res = secp256k1.verify(signature, messageHash, pubKey);
  return res;
};

const generateUUID = () => {
  return  uuidv4();
};

const associate = (primarySignature, primaryMessage, primaryPublicKey, secondarySignature, secondaryMessage, secondaryPublicKey) => {
  return (verifySignature(primarySignature, primaryMessage, primaryPublicKey) && verifySignature(secondarySignature, secondaryMessage, secondaryPublicKey)); 
};

const sessionless = {
  generateKeys,
  getKeys,
  sign,
  verifySignature,
  generateUUID,
  associate
};

export default sessionless;

File: src/haskell/lib/sessionless.hs
import Text.Printf (printf)

data KeyTuple = KeyTuple {privateKey :: String, publicKey :: String}

generateKeys :: IO (KeyTuple -> ())  -> IO (() -> ()) -> KeyTuple 
generateKeys f g = do
  putStrLn "generateKeys" 
  let kt = KeyTuple {privateKey = "foo", publicKey = "bar"}
  return kt


sign :: String -> IO String
sign message = do 
  putStrLn "message"
  return message

verifySignature :: String -> String -> String -> IO Bool
verifySignature message signature pubKey = do 
  printf "verify signature of %s %s %s\n" message signature pubKey
  return True

generateUUID :: () -> IO String
generateUUID _ = do
  putStrLn "generate uuid here"
  return "uuid"

associateKeys :: String -> String -> String -> String -> String -> String -> IO Bool
associateKeys message signature pubKey message2 signature2 pubKey2 = do 
  printf "verify signature one %s %s %s and two %s %s %s\n" message signature pubKey message2 signature2 pubKey2
  return True


main :: IO ()
main = do
  -- so what should happen here for actually calling these functions?

  generateKeys (\keyTuple -> putStrLn "save keys") (return ())
--  generateKeys (kt -> ())  ()

  sign "foo"

  (verifySignature "foo" "sig" "pubKey")

  generateUUID ()

--  putStrLn associateKeys "foo" "sig" "pubKey" "foo" "sig2" "pubKey2"


